src/s2e/Plugins/Models/CRC.cpp

```diff
@@ -0,0 +1,163 @@
+}
+
+ref<Expr> BaseFunctionModels::crc32(const ref<Expr> &initialCrc, const std::vector<ref<Expr>> &input, bool xorResult) {
+    assert(initialCrc->getWidth() == Expr::Int32);
+    ref<Expr> result = XorExpr::create(initialCrc, E_CONST(0xFFFFFFFF, Expr::Int32));
+
+    for (uint32_t i = 0; i < input.size(); i++) {
+        ref<Expr> c = input[i];
+        assert(c->getWidth() == Expr::Int8);
+
+        ref<Expr> crc32_low8 = E_EXTR(result, 0, Expr::Int8);
+
+        // index = ((result & 0xFF) ^ c) << 2
+        ref<Expr> index = ShlExpr::create(E_ZE(XorExpr::create(crc32_low8, c), Expr::Int32), E_CONST(2, Expr::Int32));
+        ref<Expr> lookup;
+
+        // TODO: assumes little-endian
+        if (isa<ConstantExpr>(index)) {
+            ref<ConstantExpr> ci = dyn_cast<ConstantExpr>(index);
+            lookup = E_CONST(s_crc32[ci->getZExtValue() >> 2], Expr::Int32);
+        } else {
+            lookup = ConcatExpr::create(ReadExpr::create(*m_crc32_ul, AddExpr::create(index, E_CONST(3, Expr::Int32))),
+                                        ReadExpr::create(*m_crc32_ul, AddExpr::create(index, E_CONST(2, Expr::Int32))));
+            lookup = ConcatExpr::create(lookup,
+                                        ReadExpr::create(*m_crc32_ul, AddExpr::create(index, E_CONST(1, Expr::Int32))));
+            lookup = ConcatExpr::create(lookup, ReadExpr::create(*m_crc32_ul, index));
+        }
+
+        ref<Expr> crc32_shr8 = LShrExpr::create(result, E_CONST(8, Expr::Int32));
+        result = XorExpr::create(lookup, crc32_shr8);
+    }
+
+    // Most CRC implementations xor the final result, some don't
+    if (xorResult) {
+        result = XorExpr::create(result, E_CONST(0xFFFFFFFF, Expr::Int32));
+    }
+
+    return result;
+}
+
+ref<Expr> BaseFunctionModels::crc16(const ref<Expr> &initialCrc, const std::vector<ref<Expr>> &input) {
+    assert(initialCrc->getWidth() == Expr::Int16);
+    ref<Expr> result = XorExpr::create(initialCrc, E_CONST(0xFFFF, Expr::Int16));
+
+    for (uint32_t i = 0; i < input.size(); i++) {
+        ref<Expr> c = input[i];
+        assert(c->getWidth() == Expr::Int8);
+
+        ref<Expr> crc32_lo8 = E_EXTR(result, 0, Expr::Int8);
+        ref<Expr> crc32_hi8 = LShrExpr::create(result, E_CONST(8, Expr::Int16));
+
+        ref<Expr> index = ShlExpr::create(E_ZE(XorExpr::create(crc32_lo8, c), Expr::Int32), E_CONST(1, Expr::Int32));
+
+        // TODO: assumes little-endian
+        ref<Expr> lookup;
+        if (isa<ConstantExpr>(index)) {
+            ref<ConstantExpr> ci = dyn_cast<ConstantExpr>(index);
+            lookup = E_CONST(s_crc16[ci->getZExtValue() >> 1], Expr::Int16);
+        } else {
+            lookup = ConcatExpr::create(ReadExpr::create(*m_crc16_ul, AddExpr::create(index, E_CONST(1, Expr::Int32))),
+                                        ReadExpr::create(*m_crc16_ul, index));
+        }
+
+        result = XorExpr::create(lookup, crc32_hi8);
+    }
+
+    return result;
+}
```

Comment:
Maybe add the namespaces here in a comment, similar to StaticFunctionModels.cpp. It makes it easier to read all those braces